# PyCity Schools Analysis

Well done! Having spent years analyzing financial records for big banks, you've finally scratched your idealistic itch and joined the education sector. In your latest role, you've become the Chief Data Scientist for your city's school district. In this capacity, you'll be helping the  school board and mayor make strategic decisions regarding future school budgets and priorities.

As a first task, you've been asked to analyze the district-wide standardized test results. You'll be given access to every student's math and reading scores, as well as various information on the schools they attend. Your responsibility is to aggregate the data to and showcase obvious trends in school performance.


### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd
import numpy as np
#display values without scientic format
pd.set_option('display.float_format', lambda x: '%.3f' % x)

# File to Load (Remember to Change These)
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

# Read School and Student Data File and store into Pandas Data Frames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

#view school data
school_data.head()

School ID            school_name      type  size   budget
0          0      Huang High School  District  2917  1910635
1          1   Figueroa High School  District  2949  1884411
2          2    Shelton High School   Charter  1761  1056600
3          3  Hernandez High School  District  4635  3022020
4          4    Griffin High School   Charter  1468   917500

In [2]:
#view student data
student_data.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  
0             66          79  
1             94          61  
2             90          60  
3             67          58  
4             97          84

In [3]:
# Combine the data into a single dataset
school_data_complete = pd.merge(student_data, school_data, how="left", on="school_name")
school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

In [4]:
#Show columns
school_data_complete.columns

Index(['Student ID', 'student_name', 'gender', 'grade', 'school_name',
       'reading_score', 'math_score', 'School ID', 'type', 'size', 'budget'],
      dtype='object')

## District Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average math score 

* Calculate the average reading score

* Calculate the overall passing rate (overall average score), i.e. (avg. math score + avg. reading score)/2

* Calculate the percentage of students with a passing math score (70 or greater)

* Calculate the percentage of students with a passing reading score (70 or greater)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting

In [5]:
#Calculate the number of unique schools in dataframe
totalSchool = len(school_data_complete['School ID'].unique())
print(f'total number of schools: {totalSchool}')
#Calculate total number of students
totalStudents = len(school_data_complete['Student ID'].unique())
print(f'total number of students: {totalStudents}')
#Calculate the total budget
totalBudget = sum(school_data_complete['budget'].unique())
print(f'total budget: {totalBudget}')
#Calculate the average math score
avgMath = round(school_data_complete['math_score'].mean(),2)
print(f'average math score: {avgMath}')
#Calculate the average reading score
avgReading = round(school_data_complete['reading_score'].mean(),2)
print(f'average reading score: {avgReading}')
#Calculate the overall passing rate 
#(overall average score), i.e. (avg. math score + avg. reading score)/2
overallPass = round((avgMath + avgReading)/2,2)
print(f'overall passing rate (average of math and reading): {overallPass}')
#Calculate the percentage of students with a passing math score (70 or greater)
mathPass = round(100*(school_data_complete['Student ID'][school_data_complete['math_score']>=70].count())/
                 totalStudents,2)
print(f'percentage of students with a passing math score 70 or greater: {mathPass}')
#Calculate the percentage of students with a passing reading score (70 or greater)
readingPass = round(100*(school_data_complete['Student ID'][school_data_complete['reading_score']>=70].count())/
                    totalStudents,2)
print(f'percentage of students with passing reading score 70 or greater: {readingPass}')

total number of schools: 15
total number of students: 39170
total budget: 24649428
average math score: 78.99
average reading score: 81.88
overall passing rate (average of math and reading): 80.44
percentage of students with a passing math score 70 or greater: 74.98
percentage of students with passing reading score 70 or greater: 85.81


In [6]:
#Create a dataframe to hold the above results
school_summary = pd.DataFrame({
    "Total Schools": [totalSchool],
    "Total Students": [totalStudents],
    "Total Budget": [totalBudget],
    "Average Math Score": [avgMath],
    "Average Reading Score": [avgReading],
    "% Passing Math": [mathPass],
    "% Passing Reading": [readingPass],
    "% Overall Passing Rate": [overallPass]
})
#Optional: give the displayed data cleaner formatting
school_summary['Total Students'] = school_summary['Total Students'].map(
    "{0:,.0f}".format)
school_summary['Total Budget'] = school_summary['Total Budget'].map(
    "${0:,.2f}".format)
school_summary['% Passing Math'] = school_summary['% Passing Math'].map(
    "{0:,.2f}%".format)
school_summary['% Passing Reading'] = school_summary['% Passing Reading'].map(
    "{0:,.2f}%".format)
school_summary['% Overall Passing Rate'] = school_summary['% Overall Passing Rate'].map(
    "{0:,.2f}%".format)
school_summary

Total Schools Total Students    Total Budget  Average Math Score  \
0             15         39,170  $24,649,428.00              78.990   

   Average Reading Score % Passing Math % Passing Reading  \
0                 81.880         74.98%            85.81%   

  % Overall Passing Rate  
0                 80.44%

## School Summary

* Create an overview table that summarizes key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)
  
* Create a dataframe to hold the above results

In [7]:
#Calculate Metrics by grouping by School Name

schoolType_df = school_data[['school_name','type']].rename(columns={'type':'School Type'})
schoolGrp_df = school_data.groupby('school_name').sum()[['size','budget']].rename(columns={
    'size':'Total Students',
    'budget':'Total School Budget'
})

#Merge total size, budget, and type
schoolGrp_df = schoolGrp_df.merge(schoolType_df,on='school_name')
schoolGrp_df

school_name  Total Students  Total School Budget School Type
0      Bailey High School            4976              3124928    District
1     Cabrera High School            1858              1081356     Charter
2    Figueroa High School            2949              1884411    District
3        Ford High School            2739              1763916    District
4     Griffin High School            1468               917500     Charter
5   Hernandez High School            4635              3022020    District
6      Holden High School             427               248087     Charter
7       Huang High School            2917              1910635    District
8     Johnson High School            4761              3094650    District
9        Pena High School             962               585858     Charter
10  Rodriguez High School            3999              2547363    District
11    Shelton High School            1761              1056600     Charter
12     Thomas High School            1635              1043130     Charter
13     Wilson High School            2283              1319574     Charter
14     Wright High School            1800              1049400     Charter

In [8]:
#Calculate Metrics by grouping by School Name - for averages using mean()

#average math and reading score
avg_df = school_data_complete.groupby('school_name').mean()[['math_score','reading_score']].rename(
    columns={
        'math_score':'Average Math Score',
        'reading_score':'Average Reading Score'})

#merge school data with average scores
merge_df1 = pd.merge(schoolGrp_df,avg_df,on='school_name')

#per student budget
merge_df1['Per Student Budget'] = merge_df1['Total School Budget']/merge_df1['Total Students']
merge_df1.head()

school_name  Total Students  Total School Budget School Type  \
0    Bailey High School            4976              3124928    District   
1   Cabrera High School            1858              1081356     Charter   
2  Figueroa High School            2949              1884411    District   
3      Ford High School            2739              1763916    District   
4   Griffin High School            1468               917500     Charter   

   Average Math Score  Average Reading Score  Per Student Budget  
0              77.048                 81.034             628.000  
1              83.062                 83.976             582.000  
2              76.712                 81.158             639.000  
3              77.103                 80.746             644.000  
4              83.351                 83.817             625.000

In [24]:
#Calculate Metrics by grouping by School Name - Total Students with Passing Math and Reading Scores
mathPass_df = school_data_complete[school_data_complete['math_score']>=70][['school_name','math_score']].groupby(
    'school_name').count().rename(columns={'math_score':'Total Students with Passing Math Scores'})

readingPass_df = school_data_complete[
    school_data_complete['reading_score']>=70][['school_name','reading_score']].groupby(
    'school_name').count().rename(columns={'reading_score':'Total Students with Passing Reading Scores'})

#Merge number of students with passing scores with merge_df1
merge_df2 = merge_df1.merge(mathPass_df,on='school_name')
merge_df3 = merge_df2.merge(readingPass_df,on='school_name')

#Calculate average passing scores
merge_df3['% Passing Math'] = 100*(merge_df3['Total Students with Passing Math Scores']/
                                   merge_df3['Total Students'])

merge_df3['% Passing Reading'] = 100*(merge_df3['Total Students with Passing Reading Scores']/
                                      merge_df3['Total Students'])

merge_df3['% Overall Passing Rate'] = (merge_df3['% Passing Math'] + merge_df3['% Passing Reading'])/2
merge_df3.head()

school_name  Total Students  Total School Budget School Type  \
0    Bailey High School            4976              3124928    District   
1   Cabrera High School            1858              1081356     Charter   
2  Figueroa High School            2949              1884411    District   
3      Ford High School            2739              1763916    District   
4   Griffin High School            1468               917500     Charter   

   Average Math Score  Average Reading Score  Per Student Budget  \
0              77.048                 81.034             628.000   
1              83.062                 83.976             582.000   
2              76.712                 81.158             639.000   
3              77.103                 80.746             644.000   
4              83.351                 83.817             625.000   

   Total Students with Passing Math Scores  \
0                                     3318   
1                                     1749   
2                                     1946   
3                                     1871   
4                                     1371   

   Total Students with Passing Reading Scores  % Passing Math  \
0                                        4077          66.680   
1                                        1803          94.133   
2                                        2381          65.988   
3                                        2172          68.310   
4                                        1426          93.392   

   % Passing Reading  % Overall Passing Rate  
0             81.933                  74.307  
1             97.040                  95.587  
2             80.739                  73.364  
3             79.299                  73.804  
4             97.139                  95.266

In [10]:
#Save as new Summary DataFrame with Cleaner Formatting

schoolSummary_df = merge_df3[[
    'school_name',
    'School Type',
    'Total Students',
    'Total School Budget',
    'Per Student Budget',
    'Average Math Score',
    'Average Reading Score',
    '% Passing Math',
    '% Passing Reading',
    '% Overall Passing Rate']].set_index('school_name')

schoolSummary_df['Total School Budget'] = schoolSummary_df['Total School Budget'].map(
    "${0:,.2f}".format)
schoolSummary_df['Per Student Budget'] = schoolSummary_df['Per Student Budget'].map(
    "${0:,.2f}".format)
schoolSummary_df['% Passing Math'] = schoolSummary_df['% Passing Math'].map(
    "{0:,.2f}%".format)
schoolSummary_df['% Passing Reading'] = schoolSummary_df['% Passing Reading'].map(
    "{0:,.2f}%".format)
schoolSummary_df['% Overall Passing Rate'] = schoolSummary_df['% Overall Passing Rate'].map(
    "{0:,.2f}%".format)
schoolSummary_df

School Type  Total Students Total School Budget  \
school_name                                                             
Bailey High School       District            4976       $3,124,928.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Figueroa High School     District            2949       $1,884,411.00   
Ford High School         District            2739       $1,763,916.00   
Griffin High School       Charter            1468         $917,500.00   
Hernandez High School    District            4635       $3,022,020.00   
Holden High School        Charter             427         $248,087.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Pena High School          Charter             962         $585,858.00   
Rodriguez High School    District            3999       $2,547,363.00   
Shelton High School       Charter            1761       $1,056,600.00   
Thomas High School        Charter            1635       $1,043,130.00   
Wilson High School        Charter            2283       $1,319,574.00   
Wright High School        Charter            1800       $1,049,400.00   

                      Per Student Budget  Average Math Score  \
school_name                                                    
Bailey High School               $628.00              77.048   
Cabrera High School              $582.00              83.062   
Figueroa High School             $639.00              76.712   
Ford High School                 $644.00              77.103   
Griffin High School              $625.00              83.351   
Hernandez High School            $652.00              77.290   
Holden High School               $581.00              83.803   
Huang High School                $655.00              76.629   
Johnson High School              $650.00              77.072   
Pena High School                 $609.00              83.840   
Rodriguez High School            $637.00              76.843   
Shelton High School              $600.00              83.359   
Thomas High School               $638.00              83.418   
Wilson High School               $578.00              83.274   
Wright High School               $583.00              83.682   

                       Average Reading Score % Passing Math % Passing Reading  \
school_name                                                                     
Bailey High School                    81.034         66.68%            81.93%   
Cabrera High School                   83.976         94.13%            97.04%   
Figueroa High School                  81.158         65.99%            80.74%   
Ford High School                      80.746         68.31%            79.30%   
Griffin High School                   83.817         93.39%            97.14%   
Hernandez High School                 80.934         66.75%            80.86%   
Holden High School                    83.815         92.51%            96.25%   
Huang High School                     81.183         65.68%            81.32%   
Johnson High School                   80.966         66.06%            81.22%   
Pena High School                      84.045         94.59%            95.95%   
Rodriguez High School                 80.745         66.37%            80.22%   
Shelton High School                   83.726         93.87%            95.85%   
Thomas High School                    83.849         93.27%            97.31%   
Wilson High School                    83.989         93.87%            96.54%   
Wright High School                    83.955         93.33%            96.61%   

                      % Overall Passing Rate  
school_name                                   
Bailey High School                    74.31%  
Cabrera High School                   95.59%  
Figueroa High School                  73.36%  
Ford High School                      73.80%  
Griffin High School            

* Sort and display the top five schools in overall passing rate

In [11]:
#Top 5 performing schools
schoolSummary_df.sort_values('% Overall Passing Rate',ascending=False).head()

School Type  Total Students Total School Budget  \
school_name                                                           
Cabrera High School     Charter            1858       $1,081,356.00   
Thomas High School      Charter            1635       $1,043,130.00   
Griffin High School     Charter            1468         $917,500.00   
Pena High School        Charter             962         $585,858.00   
Wilson High School      Charter            2283       $1,319,574.00   

                    Per Student Budget  Average Math Score  \
school_name                                                  
Cabrera High School            $582.00              83.062   
Thomas High School             $638.00              83.418   
Griffin High School            $625.00              83.351   
Pena High School               $609.00              83.840   
Wilson High School             $578.00              83.274   

                     Average Reading Score % Passing Math % Passing Reading  \
school_name                                                                   
Cabrera High School                 83.976         94.13%            97.04%   
Thomas High School                  83.849         93.27%            97.31%   
Griffin High School                 83.817         93.39%            97.14%   
Pena High School                    84.045         94.59%            95.95%   
Wilson High School                  83.989         93.87%            96.54%   

                    % Overall Passing Rate  
school_name                                 
Cabrera High School                 95.59%  
Thomas High School                  95.29%  
Griffin High School                 95.27%  
Pena High School                    95.27%  
Wilson High School                  95.20%

## Bottom Performing Schools (By Passing Rate)

* Sort and display the five worst-performing schools

In [12]:
#Bottom 5 performing schools
schoolSummary_df.sort_values('% Overall Passing Rate').head()

School Type  Total Students Total School Budget  \
school_name                                                             
Rodriguez High School    District            3999       $2,547,363.00   
Figueroa High School     District            2949       $1,884,411.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Ford High School         District            2739       $1,763,916.00   

                      Per Student Budget  Average Math Score  \
school_name                                                    
Rodriguez High School            $637.00              76.843   
Figueroa High School             $639.00              76.712   
Huang High School                $655.00              76.629   
Johnson High School              $650.00              77.072   
Ford High School                 $644.00              77.103   

                       Average Reading Score % Passing Math % Passing Reading  \
school_name                                                                     
Rodriguez High School                 80.745         66.37%            80.22%   
Figueroa High School                  81.158         65.99%            80.74%   
Huang High School                     81.183         65.68%            81.32%   
Johnson High School                   80.966         66.06%            81.22%   
Ford High School                      80.746         68.31%            79.30%   

                      % Overall Passing Rate  
school_name                                   
Rodriguez High School                 73.29%  
Figueroa High School                  73.36%  
Huang High School                     73.50%  
Johnson High School                   73.64%  
Ford High School                      73.80%

## Math Scores by Grade

* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

  * Create a pandas series for each grade. Hint: use a conditional statement.
  
  * Group each series by school
  
  * Combine the series into a dataframe
  
  * Optional: give the displayed data cleaner formatting

In [13]:
#Columns as 9th, 10th, 11th, and 12th grade, Row as Schools, Values as average math and reading scores

#Pandas series for each grade 
grade_9 = school_data_complete[school_data_complete['grade']=='9th'].groupby('school_name')['math_score'].mean()
grade_10 = school_data_complete[school_data_complete['grade']=='10th'].groupby('school_name')['math_score'].mean()
grade_11 = school_data_complete[school_data_complete['grade']=='11th'].groupby('school_name')['math_score'].mean()
grade_12 = school_data_complete[school_data_complete['grade']=='12th'].groupby('school_name')['math_score'].mean()

#Series into dataframe
mathScores_df = pd.DataFrame({
    '9th':grade_9,
    '10th':grade_10,
    '11th':grade_11,
    '12th':grade_12
})
mathScores_df

#Cleaner formatting
mathScores_df.style.format({ 
    '9th':  '{:.2f}', 
    '10th': '{:.2f}', 
    '11th': '{:.2f}', 
    '12th': '{:.2f}'})

## Reading Score by Grade 

* Perform the same operations as above for reading scores

In [14]:
#Columns as 9th, 10th, 11th, and 12th grade, Row as Schools, Values as average reading and reading scores

#Pandas series for each grade
grade_9 = school_data_complete[school_data_complete['grade']=='9th'].groupby('school_name')['reading_score'].mean()
grade_10 = school_data_complete[school_data_complete['grade']=='10th'].groupby('school_name')['reading_score'].mean()
grade_11 = school_data_complete[school_data_complete['grade']=='11th'].groupby('school_name')['reading_score'].mean()
grade_12 = school_data_complete[school_data_complete['grade']=='12th'].groupby('school_name')['reading_score'].mean()

#No brackets for series into dataframe
readingScores_df = pd.DataFrame({
    '9th':grade_9,
    '10th':grade_10,
    '11th':grade_11,
    '12th':grade_12
})
readingScores_df

#Cleaner formatting
readingScores_df.style.format({ 
    '9th':  '{:.2f}', 
    '10th': '{:.2f}', 
    '11th': '{:.2f}', 
    '12th': '{:.2f}'})

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [15]:
#Create a new School Summary without formatting
schoolSummary_df2 = merge_df3[[
    'school_name',
    'School Type',
    'Total Students',
    'Total School Budget',
    'Per Student Budget',
    'Average Math Score',
    'Average Reading Score',
    '% Passing Math',
    '% Passing Reading',
    '% Overall Passing Rate']].set_index('school_name')
schoolScores = schoolSummary_df2.reset_index()
schoolScores
#View summary statistics for schoolSummary data
schoolScores.describe()

Total Students  Total School Budget  Per Student Budget  \
count          15.000               15.000              15.000   
mean         2611.333          1643295.200             620.067   
std          1420.915           934776.294              28.544   
min           427.000           248087.000             578.000   
25%          1698.000          1046265.000             591.500   
50%          2283.000          1319574.000             628.000   
75%          3474.000          2228999.000             641.500   
max          4976.000          3124928.000             655.000   

       Average Math Score  Average Reading Score  % Passing Math  \
count              15.000                 15.000          15.000   
mean               80.433                 82.529          80.987   
std                 3.375                  1.519          13.999   
min                76.629                 80.745          65.684   
25%                77.060                 81.000          66.523   
50%                83.062                 83.726          92.506   
75%                83.389                 83.902          93.630   
max                83.840                 84.045          94.595   

       % Passing Reading  % Overall Passing Rate  
count             15.000                  15.000  
mean              89.219                  85.103  
std                8.181                  11.072  
min               79.299                  73.293  
25%               81.043                  73.722  
50%               95.855                  94.379  
75%               96.575                  95.235  
max               97.309                  95.587

In [16]:
#Create bins by quartiles

#Spending budget per student
spending_bins = [0, 592, 629, 642, 656]
spending_labels = ["(bottom 25th) $592", "(25th to 50th) $592-<629", "(50th to 75th) $629-<642", "(75th+) $642-655"]

In [17]:
#Add spending ranges
schoolScores["Spending Ranges Per Student"] = pd.cut(
    schoolScores["Per Student Budget"],bins=spending_bins,labels=spending_labels,right=False)
spendingRange = schoolScores.groupby('Spending Ranges Per Student').mean()[[
    'Average Math Score','Average Reading Score','% Passing Math','% Passing Reading','% Overall Passing Rate']]
spendingRange

#Cleaner formatting
spendingRange.style.format({ 
    '% Passing Math':  '{0:,.2f}%', 
    '% Passing Reading': '{0:,.2f}%', 
    '% Overall Passing Rate': '{0:,.2f}%'
})


## Scores by School Size

* Perform the same operations as above, based on school size.

In [18]:
#Create bins by quartiles
#School size
size_bins = [0, 1699, 2284, 3475, 4977]
size_labels = ["(bottom 25th) 0-<1699", "(25th to 50th) 1699-<2284", "(50th to 75th) 2284-<3575", "(75th+) 3475-4977"]

In [19]:
#Add size ranges
schoolScores["School Size"] = pd.cut(schoolScores["Total Students"],bins=size_bins,labels=size_labels,right=False)
schoolSize = schoolScores.groupby('School Size').mean()[[
    'Average Math Score','Average Reading Score','% Passing Math','% Passing Reading','% Overall Passing Rate']]

#Cleaner formatting
schoolSize.style.format({ 
    '% Passing Math':  '{0:,.2f}%', 
    '% Passing Reading': '{0:,.2f}%', 
    '% Overall Passing Rate': '{0:,.2f}%'
})

## Scores by School Type

* Perform the same operations as above, based on school type.

In [20]:
#Scores by School Type
schoolType = schoolScores.groupby('School Type').mean()[[
    'Average Math Score','Average Reading Score','% Passing Math','% Passing Reading','% Overall Passing Rate']]
#Cleaner formatting
schoolType.style.format({ 
    '% Passing Math':  '{0:,.2f}%', 
    '% Passing Reading': '{0:,.2f}%', 
    '% Overall Passing Rate': '{0:,.2f}%'
})

In [21]:
#Scores by School Type and Size
schoolScores['School Size'] = pd.cut(schoolScores["Total Students"],bins=size_bins,labels=size_labels,right=False)
schoolTypeSize = schoolScores.groupby(['School Type','School Size']).mean()[[
    'Average Math Score','Average Reading Score','% Passing Math','% Passing Reading','% Overall Passing Rate']]

#Cleaner formatting
schoolTypeSize.style.format({ 
    '% Passing Math':  '{0:,.2f}%', 
    '% Passing Reading': '{0:,.2f}%', 
    '% Overall Passing Rate': '{0:,.2f}%'
})

In [22]:
#Scores by School Type and Budget
schoolScores["Spending Ranges Per Student"] = pd.cut(
    schoolScores["Per Student Budget"],bins=spending_bins,labels=spending_labels,right=False)
spendingTypeRange = schoolScores.groupby(['School Type','school_name','Spending Ranges Per Student']).mean()[[
    'Total Students',
    'Average Math Score',
    'Average Reading Score',
    '% Passing Math',
    '% Passing Reading',
    '% Overall Passing Rate']]
spendingTypeRange.sort_values('% Overall Passing Rate')

#Cleaner formatting
spendingTypeRange.style.format({ 
    '% Passing Math':  '{0:,.2f}%', 
    '% Passing Reading': '{0:,.2f}%', 
    '% Overall Passing Rate': '{0:,.2f}%'
})

# PyCity Schools Analysis - Observable Trends in Data

---- 

* As a whole, schools with higher budgets, did not yield better test results. By contrast, schools with higher spending per student actually (\$645-675) underperformed compared to schools with smaller budgets (<\$585 per student).

* As a whole, smaller and medium sized schools dramatically out-performed large sized schools on passing math performances (89-91% passing vs 67%).

* As a whole, charter schools out-performed the public district schools across all metrics. However, more analysis will be required to glean if the effect is due to school practices or the fact that charter schools tend to serve smaller student populations per school. 
---

* 1: Charter schools have student population that is in the 50th percentile and below as compared to the entire population, while District schools have a much larger student population with size 50th percentile and above. At first glance of the data with only stratifying by School Type, it seems that 'Charter' schools out-perform District schools. However, when adding another stratification 'Total Students', the size of school confounds this fact. 

* 2: Spending may not be the major contributor for increasing math and reading scores in schools. Charter schools spend less on per student then District schools, while Charter school outperform District schools on all math and reading averages as well as passing rates. In order to increase school performance in District schools, consider a threshold of student body size in one school or increase teacher to student ratio in classrooms.

* 3: Cabrera High School has the highest '% Overall Passing Rate' (95.59%) while it's 'Per Student Spending' is in the bottom 25th percentile (\$582). This school should be monitored closely as a model for other schools wanting to improve math and reading passing rate. 